[OpenApi] Generate schema for JSON Patch endpoints

martincostello opened the pull request:
# Generate schema for JSON Patch endpoints

Generate an OpenAPI schema for JSON Patch endpoints.

## Description

Follow-up for https://github.com/dotnet/aspnetcore/pull/62988#discussion_r2246474495.

I did a bit Googling and came up with a schema that's a synthesis of different examples I found. It could potentially be simplified to not use `OneOf`.

I couldn't see any obvious reason to generate a different schema for `JsonPatchDocument<T>` compared to `JsonPatchDocument` as the patch schema is the same for both as it isn't strongly-typed to the document being patched, so both use the same schema.

Tested locally by adding SwaggerUI to the Sample App and viewing the operation generated, then using the sample request as-is and submitting it and inspecting the in-memory document in the debugger to check that it was deserialized correctly and all seems to work as expected.

<img width="1443" height="1214" alt="image" src="https://github.com/user-attachments/assets/c18b7430-1628-4b7d-bc76-72a8f4954cbd" />

martincostello committed:
[OpenApi] Add non-generic JSON patch

Update the sample to include a non-generic JSON patch endpoint.

martincostello committed:
[OpenApi] Generate schema for JSON Patch

Generate an appropriate OpenAPI schema for JSON Patch endpoints.

martincostello committed:
Fix condition

Leftovers from before tests were added.

martincostello committed:
Handle derived types

- Handle custom types derived from `JsonPatchDocument` or `JsonPatchDocument<T>`.
- Add extension method to remove duplicated type checks.

martincostello commented:
Looks like adding `Microsoft.AspNetCore.JsonPatch.SystemTextJson` as a dependency of `Microsoft.AspNetCore.OpenApi` causes trim warnings as the former is not AoT compatible but the latter is.

Having a quick look at JsonPatch there's a tonne of `object` in there and I don't think it can be made trivially AoT compatible (it would probably need a source generator?).

A workaround might be to remove the dependency and instead check on the type name, even though that's ick.

martincostello committed:
[OpenApi] Remove JsonPatch.SystemTextJson dep

Remove the dependency on `Microsoft.AspNetCore.JsonPatch.SystemTextJson` as it creates type warnings, and instead check the types by name.

captainsafia commented:
> Looks like adding Microsoft.AspNetCore.JsonPatch.SystemTextJson as a dependency of Microsoft.AspNetCore.OpenApi causes trim warnings as the former is not AoT compatible but the latter is.

Yep, I figured this problem would come up. 😅 I think the workaround of a type name + assembly name check is probably sufficient for now. In an ideal world, the M.A.JP.STJ package would've been trim-friendly to start but that didn't end up being the chase. :/

captainsafia commented on src/OpenApi/src/Extensions/TypeExtensions.cs:
```diff
@@ -0,0 +1,32 @@
+        while (modelType != null && modelType != typeof(object))
+        {
+            if (modelType.Namespace == JsonPatchDocumentNamespace &&
+                (modelType.Name == JsonPatchDocumentName || modelType.Name.StartsWith(JsonPatchDocumentNameOfT, StringComparison.Ordinal)))
```
Nit: Check if the type is generic before doing the `StartsWith` comparison?

Maybe add a check on the number of generic args being 1 as well?

captainsafia commented on src/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests/Integration/snapshots/OpenApiDocumentLocalizationTests.VerifyOpenApiDocumentIsInvariant.verified.txt:
```diff
@@ -1388,6 +1432,76 @@
+              "additionalProperties": false
+            },
+            {
+              "type": "object",
```
It seems like we can do `required: ["op", "path"]` for a remove based on https://datatracker.ietf.org/doc/html/rfc6902/#appendix-A.3?

martincostello replied:
👀 - thanks, I didn't think to check for an RFC.

captainsafia reviewed (commented):
Looks good overall! Left some small comments about a tweak to the schema. Would also like @mikekistler to chime in on the proposed schema here.

martincostello committed:
Update type check

- Check for a generic type before type name check.
- Replace `StartsWith()` with `==`.

martincostello committed:
Add required for JSON Patch remove

Add `required` members to the OpenAPI schema for a JSON Patch remove operation.

mikekistler reviewed (commented):
This looks great! Thanks @martincostello !
